src/main/java/frc/robot/subsystems/shooterLimelight/ShooterLimelight.java

```diff
@@ -0,0 +1,404 @@
+    kCameraHeight = heightMeters;
+    kTargetHeight = targetHeightMeters;
+    setLimelightValues(
+        Constants.limelightLedMode.OFF,
+        Constants.limelightCamMode.VISION_PROCESSING,
+        Constants.limelightPipeline.TAG_PIPELINE);
+
+    m_simDevice = SimDevice.create(ShooterLimelightConstants.SHOOTER_LIMELIGHT_NAME);
+    if (m_simDevice != null) {
+      m_targetArea = m_simDevice.createDouble("Target Area", Direction.kBidir, 0.0);
+      m_skew = m_simDevice.createDouble("Skew", Direction.kBidir, 0.0);
+      m_latency = m_simDevice.createDouble("Latency", Direction.kBidir, 0.0);
+      m_tx = m_simDevice.createDouble("Tx", Direction.kBidir, 0.0);
+      m_ty = m_simDevice.createDouble("Ty", Direction.kBidir, 0.0);
+      m_valid = m_simDevice.createBoolean("Valid", Direction.kBidir, false);
+    }
+  }
+
+  private static Transform2d getBotFromTarget(Pose3d botPoseTargetSpace) {
+    /**
+     * Target space: 3d Cartesian Coordinate System with (0,0,0) at the center of the target.
+     *
+     * <p>X+ → Pointing to the right of the target (If you are looking at the target)
+     *
+     * <p>Y+ → Pointing downward
+     *
+     * <p>Z+ → Pointing out of the target (orthogonal to target’s plane).
+     */
+
+    /**
+     * We convert to 2d target space: X+ -> Out of the target Y+ -> Pointing to the right of the
+     * target (If you are looking at the target) This means positive yaw is based on Z+ being up
+     */
+    Translation2d translation =
+        new Translation2d(botPoseTargetSpace.getZ(), botPoseTargetSpace.getX());
+    Rotation2d rot = Rotation2d.fromDegrees(-botPoseTargetSpace.getRotation().getY());
+
+    System.out.println("Tag at " + -botPoseTargetSpace.getRotation().getY() + " deg");
+    return new Transform2d(translation, rot);
+  }
+
+  private static boolean validScoringTag(double tagId) {
+    long tagIdRounded = Math.round(tagId);
+    return (tagIdRounded == 3 || tagIdRounded == 4 || tagIdRounded == 7 || tagIdRounded == 8);
+  }
+
+  public static int chooseTag(LimelightTarget_Fiducial[] targets) {
+    int numTags = targets.length;
+    double minDistMeters = Double.MAX_VALUE;
+    int bestTag = 0;
+    for (int tagIndex = 0; tagIndex < numTags; tagIndex++) {
+      LimelightTarget_Fiducial target = targets[tagIndex];
+      if (!validScoringTag(target.fiducialID)) {
+        continue;
+      }
+      double targetDistance = target.getTargetPose_RobotSpace().getTranslation().getNorm();
+      if (targetDistance < minDistMeters) {
+        minDistMeters = targetDistance;
+        bestTag = tagIndex;
+      }
+    }
+    return bestTag;
+  }
+
+  private Transform2d getRobotToScoringLocation(Pose3d targetPoseRobotSpace) {
+    Transform2d targetFromBot = getBotFromTarget(targetPoseRobotSpace);
+    return targetFromBot;
+  }
+
+  public Optional<Transform2d> getRobotToScoringLocation() {
+    return robotToScoringLocation;
+  }
+
+  public Optional<Transform2d> checkForTag() {
```

Comment:
I don't think this will work consistently. The "side" speaker tag might be closer and it would be consistently detected over the other one. We should use the megatag pose and use that to point at the speaker

Reply:
Or you can iterate through the tag detections and use the center one

Reply:
What is the "megatag pose"?

Reply:
Reworked in `basel/integration-v4`